design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,352 @@
+allow the key column in the projection.
+
+### Simple query
+
+The difference in key column handling can be seen with a simple query:
+
+```sql
+SELECT ID, V0, V1 FROM INPUT;
+-- resulting columns: ID INT, V0 INT, V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS SELECT ID, V0, V1 FROM INPUT;
+-- fails with error about duplicate 'ID' column.
+```
+
+Where as the above transient query works and returns all columns, converting it to a persistent
+query causes it to fail with a duplicate column error, because `ID` is _implicitly_ being copied to
+the OUTPUT schema, and the `ID` in the projection is creating a value column called `ID`. That's two
+columns named `ID` and hence the query fails.
+
+We propose that the above `CREATE TABLE` statement should work, as the projection is equivalent to
+`select *`, which works.
+
+### Select star
+
+The differences with a 'select *' projection are more subtle.
+
+For non-join queries, both transient and persistent queries select all the columns in the schema:
+
+```sql
+SELECT * FROM INPUT;
+-- resulting columns: ID INT, V0 INT, V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS SELECT * FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT (Same as above)
+```
+
+With a join query, a transient query selects all columns from all sources. A persistent query adds
+all columns from all sources, but also adds an additional copy of the left key column.
+
+```sql
+SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting columns: I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting schema: ID INT KEY, I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+```
+
+Note the addition of an additional `ID` column in the case of the persistent query.
+
+We propose that the join should not duplicate the left join column `I1.ID` into both the `ID` key
+and `I1_ID` value column.
+
+Only the key column should exist and it should be called either (TBD):
+  - `ID`, or
+  - `I1_ID`.
+
+### Removal of non-standard Aliasing
+
+The [allow any key name](https://github.com/confluentinc/ksql/issues/3536) feature has introduced
+the ability to provide an alias within a `PARTITION BY`, `GROUP BY` and `JOIN` clause. These are
+non-standard, but required in the current model to allow users to name the key column in the schema.
+For example:
+
+```sql
+-- without alias:
+CREATE TABLE OUTPUT AS SELECT ID FROM INPUT PARTITION BY V0 - V1;
+-- resulting schema: KSQL_COL_0 INT KEY, ID;
+-- note the system generated column name.
+
+-- with alias
+CREATE TABLE OUTPUT AS SELECT ID FROM INPUT PARTITION BY V0 - V1 AS NEW_KEY;
+-- resulting schema: NEW_KEY INT KEY, ID;
+```
+
+However, the same functionality can be achieved using standard sql if the key column is required
+in the projection, for example:
+
+```sql
+CREATE TABLE OUTPUT AS SELECT V0 - V1 AS NEW_KEY, ID FROM INPUT PARTITION BY V0 - V1;
```

Comment:
Above above for the `GROUP BY` example: the `PARTITION BY` is applied to `INPUT` _before_ the projection. Hence `NEW_KEY` wouldn't exist and it would fail with an unknown column.

Reply:
Same as my [previous comment](https://github.com/confluentinc/ksql/pull/5115#discussion_r414090877). I think it would be a more intuitive UX to simply swap in the expression associated with an alias so that users can do what @rmoff asked about.

Reply:
I think we don't support `PARTITION BY` on tables anyway? The PK does define the partitioning and we cannot just set a new PK via `PARTITION BY`. Hence, the example seems incorrect?

Reply:
> Same as my [previous comment](https://github.com/confluentinc/ksql/pull/5115#discussion_r414090877). I think it would be a more intuitive UX to simply swap in the expression associated with an alias so that users can do what @rmoff asked about.

I actually think this is counter intuitive, better this works in the same way as GROUP BY, but that's just IMHO.  

In SQL you can't do:

```sql
CREATE TABLE OUTPUT AS 
   SELECT V0 - V1 AS NEW_KEY, ID FROM INPUT GROUP BY NEW_KEY;
```

So it follows you shouldn't be able to do:

```sql
CREATE STREAM OUTPUT AS 
   SELECT V0 - V1 AS NEW_KEY, ID FROM INPUT PARTITION BY NEW_KEY;
```

I don't think we should change this without knowing this is what users want and the ROI is there.  We'd need a KLIP for sure.

Reply:
> I think we don't support `PARTITION BY` on tables anyway? The PK does define the partitioning and we cannot just set a new PK via `PARTITION BY`. Hence, the example seems incorrect?

nit!  but yes, this is true, it should be a stream